### PR TITLE
Piwik.js: add callback ability to trackPageView

### PR DESCRIPTION
**Pull request by er314**
because I need it :-)
and because it's so easy to do, thanks to piwik architecture :-)

exact same principles as callback for trackLink

### CODE REVIEW

File: js/piwik.js [hunk 1]
```diff
@@ -4032,10 +4032,10 @@ if (typeof window.Piwik !== 'object') {
             /*
              * Log the page view / visit
              */
-            function logPageView(customTitle, customData) {
+            function logPageView(customTitle, customData, callback) {
                 var request = getRequest('action_name=' + encodeWrapper(titleFixup(customTitle || configTitle)), customData, 'log');
 
-                sendRequest(request, configTrackerPause);
+                sendRequest(request, (callback ? 0 : configTrackerPause), callback);
```
Comment (tsteur) [inline js/piwik.js]: Thx for sending the PR 👍 

I dont really understand this part: `(callback ? 0 : configTrackerPause)`. Shouldn't we still always send `configTrackerPause` as second parameter?

**Comment by er314** [comment]
Hi, 

If I remember correctly, as can be verified in the callback code for trackLink which does exactly the same, here is the reason : 
param1 of sendRequest is the request
param2 is the wait delay before resuming
param3 is the optional callback
So, the idea is that, if a callback function is provided, then we set delay to 0, there's no delay because everything is done asynchronously and event-triggered. That's the beauty of it :-)

**Comment by er314** [comment]
Below is the line of code with same purpose, which is invoked by trackLink ; as you can see, in my current pull request for trackPageView, I did strictly copycat :-)
https://github.com/piwik/piwik/blob/master/js/piwik.js#L4590

I think the main question with this pull request, it due to adding a third, optional, argument to trackPageView. There again, I did study what was done for trackLink when the callback feature was added, and I applied the same principles. 
But I'm not expert enough in PHP to garantee that this will have zero side effects.
I can say that things work very well on my test website, that's for sure :-)

cheers

File: js/piwik.js (hunk 1 above)
Comment (tsteur) [inline js/piwik.js]: @mattab do you remember why this was needed/added in https://github.com/piwik/piwik/pull/5850 ?

**Comment by tsteur** [comment]
Thanks. It was added in https://github.com/piwik/piwik/pull/5850 but I'm not sure why and not sure if it's needed there at all. I hope someone can explain :)

File: js/piwik.js (hunk 1 above)
Comment (tsteur) [inline js/piwik.js]: @mattab please ignore it, I got it.

@er314 can you change it to pass always `configTrackerPause` as 2nd parameter?

**Comment by tsteur** [comment]
Actually I get it now why it was done initially (the `(callback ? 0 : configTrackerPause)`) but think it does not really apply here and it should be actually changed in the other `logLink` method as well to always pass `configTrackerPause`.

This condition was added that time in case someone passes this callback in which this behaviour `(callback ? 0 : configTrackerPause)` kind of made sense:

```js
_paq.push(['trackLink', url, type, {}, function() { window.location = url }]);
```

However, this is only one use case for callbacks. The callbacks can be used for many other things and when not actually using this specific callback so far, it would cause outlinks not being tracked correctly as we would never wait for the tracking request to finish when a user leaves the current website. If users want to make use of this special case they should call `tracker.setLinkTrackingTimer(0);`. I will create an issue for this. So it would be great if you could simply replace the 2nd parameter with `configTrackerPause`

**Comment by er314** [comment]
mmmm I'm not sure to understand what you mean, when you say : 

> it would cause outlinks not being tracked correctly as we would never wait for the tracking request to finish when a user leaves the current website


on the contrary, my understanding of the tracking behaviour is : the callback code will be executed when the ajax request's status will change to value 4, which means, when the ajax request is **completed**.
so, whatever the actual code of the callback, it will be executed exactly when needed ; not too early, not too late :-)
```

xhr.onreadystatechange = function () {
                        if (this.readyState === 4 && !(this.status >= 200 && this.status < 300) && fallbackToGet) {
                            getImage(request, callback);
                        } else {
                            if (this.readyState === 4 && (typeof callback === 'function')) { callback(); }
                        }
                    };
```

**Comment by tsteur** [comment]
Sure, the callback will be executed, but the tracking request might not be actually completed. When you click on a link to an external website and set `configTrackerPause = 0`, then the browser will not wait for the tracking request to finish and it might not be actually recorded etc.

**Comment by er314** [comment]
My understanding is quite the opposite : the tracking request will be completed, because the browser will execute the callback code as soon as it receives the status "ajax request completed".
Chronology : 
1. the browser binds the callback to javascript event xhr.onreadystatechange
2. the browser sends the xhr ajax request
3. once the request is complete, the javascript event xhr.onreadystatechange is triggered, with readyState value = 4
4. the callback is then executed

**Comment by tsteur** [comment]
Yes, it will execute the callback but the browser might not wait for the request to finish because it is already loading another page eg when clicking on a link. There is a callback attached to the XHR anyway but it doesn't mean the browser will wait for the XHR to finish before loading the next page/website

**Comment by er314** [comment]
understood !
you're correct. and your proposed fix + usage note makes perfect sense.
so actually thanks to this feature request we found a bug in the existing code :-)
I have just commited the change accordingly.

**Comment by tsteur** [comment]
Thank you for the quick changes 👍  It's always good to find bugs and to improve :)

**Pull request merged**